infrastructure/src/main/java/org/corfudb/infrastructure/management/HealingDetector.java

```diff
@@ -0,0 +1,233 @@
+     * Sets up the unresponsive servers array. (All servers in the unresponsiveServers list in
+     * the layout)
+     * Resets the currentEpoch to the layout epoch.
+     *
+     * @param layout       Latest known layout
+     * @param corfuRuntime Connected CorfuRuntime instance.
+     */
+    private void setup(Layout layout, CorfuRuntime corfuRuntime) {
+
+        if (currentEpoch != layout.getEpoch()) {
+            currentEpoch = layout.getEpoch();
+
+            // Get all unresponsive servers.
+            members = layout.getUnresponsiveServers().stream().toArray(String[]::new);
+            Arrays.sort(members);
+
+            log.debug("Unresponsive members to poll, {}", new ArrayList<>(Arrays.asList(members)));
+            membersRouters = new IClientRouter[members.length];
+            responses = new int[members.length];
+
+            for (int i = 0; i < members.length; i++) {
+                membersRouters[i] = corfuRuntime.getRouterFunction.apply(members[i]);
+                membersRouters[i].setTimeoutResponse(detectionPeriodDuration);
+            }
+            pollCompletableFutures = new CompletableFuture[members.length];
+
+        } else {
+            log.debug("No server list change since last poll.");
+        }
+    }
+
+    /**
+     * Poll all members servers once asynchronously and store their futures in
+     * pollCompletableFutures.
+     */
+    private void pollOnceAsync() {
+
+        // Poll servers for health.  All ping activity will happen in the background.
+        for (int i = 0; i < members.length; i++) {
+            try {
+                pollCompletableFutures[i] = membersRouters[i].getClient(BaseClient.class).ping();
+            } catch (Exception e) {
+                CompletableFuture<Boolean> cf = new CompletableFuture<>();
+                cf.completeExceptionally(e);
+                pollCompletableFutures[i] = cf;
+            }
+        }
+    }
+
+    /**
+     * Reset all responses to an invalid iteration number, -1.
+     */
+    private void resetResponses() {
+        for (int i = 0; i < responses.length; i++) {
+            responses[i] = INVALID_RESPONSE;
+        }
+    }
+
+    /**
+     * Block on all poll futures and collect the responses. There are 3 possible cases.
+     * 1. Receive a PONG. We set the responses[i] to the polling iteration number.
+     * 2. WrongEpochException is thrown. We mark as a successful response.
+     * 3. Other Exception is thrown. We do nothing. (The response[i] in this case is left behind.)
+     *
+     * @param pollIteration poll iteration in the ongoing polling round.
+     */
+    private void collectResponsesAndVerifyEpochs(int pollIteration) {
+        // Collect responses and increment response counters for successful pings.
+        for (int i = 0; i < members.length; i++) {
+            try {
+                pollCompletableFutures[i].get();
+                responses[i] = pollIteration;
+            } catch (Exception e) {
```

Comment:
log errors for exception everywhere needed. It is hard to remember but we need to make sure we have enough logging.

Reply:
I agree, but logging exceptions in this loop is really going to produce a lot of unwanted logs.
Since this is a poller fro healing nodes, it usually polls failed nodes and will encounter exceptions every second. We do not want this to pollute our debug logs.
For now, I will log when we encounter a WrongEpochException.